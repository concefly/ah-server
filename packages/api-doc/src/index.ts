import './type';

export * from './extension';
