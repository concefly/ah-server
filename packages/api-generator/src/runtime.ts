// 运行时的 util

import { IApiServiceRequest, IApiServiceMeta, IApiService } from 'ah-api-type';

export function createMethod<Q, T>(
  method: IApiServiceRequest<Q, T>,
  $meta?: IApiServiceMeta
): IApiService<Q, T> {
  return Object.assign(method, { $meta });
}

export function replaceUrlParams(url: string, data?: any) {
  if (!data) return url;

  const match = url.match(/\:\w+/g);
  if (match) {
    const pnSet = new Set(match.map(p => p.replace(/^\:/, '')));
    pnSet.forEach(p => {
      url = url.replace(new RegExp(`\:${p}`, 'g'), data[p]);
    });
  }

  return url;
}
