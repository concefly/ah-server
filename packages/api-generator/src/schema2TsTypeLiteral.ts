import { Schema } from 'ah-api-type';

export function schema2TsTypeLiteral(s?: Schema): string {
  if (!s) return 'never';

  if ('oneOf' in s) return s.oneOf.map(schema2TsTypeLiteral).join(' | ');

  if (s.type === 'string') {
    if (s.enum)
      return s.enum
        .map(t =>
          // 转换成 string 字面量
          JSON.stringify(t)
        )
        .join(' | ');
    return 'string';
  }

  if (s.type === 'integer' || s.type === 'number') {
    if (s.enum) return s.enum.join(' | ');
    return 'number';
  }

  if (s.type === 'array') {
    return `Array<${schema2TsTypeLiteral((s.items as any) || '{}')}>`;
  }

  if (s.type === 'object') {
    if (s.properties) {
      return [
        '{',
        Object.entries(s.properties).map(([pn, pv]) => {
          const isRequired =
            typeof s.required === 'boolean'
              ? s.required
              : Array.isArray(s.required)
              ? s.required.includes(pn)
              : false;

          const tsType = schema2TsTypeLiteral(pv);

          return `${pn}${isRequired ? '' : '?'}: ${tsType}`;
        }),
        '}',
      ].join(' ');
    }

    return 'object';
  }

  return 'any';
}
