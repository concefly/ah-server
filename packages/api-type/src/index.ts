export type SchemaUi = {
  list?: {
    columns?: string[];
  };
  detail?: {
    columns?: string[];
    split?: string[];
  };
  editor?: {
    columns?: string[];
  };

  hideInForm?: boolean;

  /** 表单字段定义 */
  formField?:
    | {
        type: 'select';
        /** 搜索实体 */
        entity: string;
        /** 搜索内容字段，支持 dotPath */
        queryMapper: string;
        /** 搜索结果显示字段，支持 dotPath，默认自动生成的 title */
        resultDisplayMapper?: string;
        /** 搜索结果回填表单字段，支持 dotPath，默认 id */
        fieldValueMapper?: string;
      }
    | { type: 'textarea' }
    | { type: 'never' };

  follow?: string;

  /** 渲染类型 */
  type?: 'timestamp' | 'image' | 'richtext' | 'url' | 'pre';
  renderPipe?: { type: 'truncate'; length?: number }[];

  /** 渲染的时候链接到实体 */
  linkToEntity?: {
    name: string;
    idMapper?: string;
  };
};

export type SchemaObject = {
  type: 'object';
  properties?: {
    [name: string]: Schema;
  };
  required?: string[];
};

export type SchemaArray = {
  type: 'array';
  items?: Schema;
};

export type SchemaInteger = {
  type: 'integer';
};

export type SchemaString = {
  type: 'string';
};

/** json schema 子集 */
export type Schema = (SchemaObject | SchemaArray | SchemaInteger | SchemaString) & {
  title?: string;
  description?: string;
  __ui?: SchemaUi;
};
