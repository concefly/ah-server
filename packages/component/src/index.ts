export * from './lib/CommonCRUD';
export * from './hook/useFSM';
export * from './hook/useLocalStorage';
export * from './hook/useRequest';
