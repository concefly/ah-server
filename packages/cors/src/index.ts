import { BaseExtension } from 'ah-server';
import * as cors from '@koa/cors';

export class CorsExtension extends BaseExtension {
  constructor(private opts?: cors.Options) {
    super();
  }

  getMiddleware() {
    return [cors(this.opts)];
  }
}
