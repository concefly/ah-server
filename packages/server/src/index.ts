import 'reflect-metadata';

// for type
import * as Koa from 'koa';
import 'koa-body';

import { Schema } from 'jsonschema';

export interface IApplication {}

/** 来自 app 的 service */
export interface IService {}

/** 来自 extension 的 service */
export interface IExtensionService {}
export interface IMergedService extends IService, IExtensionService {}

export interface IContext extends Koa.Context {}

export type IRouterMethod = 'GET' | 'POST' | 'DELETE' | 'PUT';
export type IMiddleware = (ctx: IContext, next: () => Promise<any>) => Promise<any>;
export interface IRouterMeta {
  path: string;
  method: IRouterMethod | IRouterMethod[];
  middlewares?: IMiddleware[];
  query?: {
    tap?: ((q: any) => any) | 'tryParseIntProperty';
    schema: Schema;
  };
}

export interface IConfig {
  LOCAL_PORT: number;
}

export * from './App';
export * from './Config';
export * from './Controller';
export * from './Service';
export * from './util';
export * from './Scheduler';
export * from './error';
export * from './Event';
export * from './Extension';
export * from './refactor';
