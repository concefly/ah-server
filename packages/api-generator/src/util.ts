import { OpenAPIV3 as API } from 'openapi-types';
import _ from 'lodash';
import { Schema } from 'ah-api-type';
import { inspect } from 'util';
import { schema2TsTypeLiteral } from './schema2TsTypeLiteral';

export const log = (...args: any[]) => {
  console.log(...args.map(arg => inspect(arg, { showHidden: false, depth: null })));
};

export function parseRequestSchema(
  data:
    | API.PathItemObject['get']
    | API.PathItemObject['put']
    | API.PathItemObject['post']
    | API.PathItemObject['delete']
): Schema | undefined {
  if (!data) return;

  const schema: Schema = {
    type: 'object',
    properties: {},
  };

  if (data.parameters) {
    (data.parameters as API.ParameterObject[]).forEach(p => {
      Object.assign(schema.properties!, {
        [p.name]: p.schema,
      });
    });
  }

  if (data.requestBody) {
    const requestBodyContentJson = (data.requestBody as API.RequestBodyObject).content[
      'application/json'
    ];

    if (typeof (requestBodyContentJson?.schema as any)?.type !== 'undefined') {
      const jSchema = requestBodyContentJson?.schema as Schema;
      _.merge(schema, jSchema);
    }
  }

  // 没有收集到，返回空
  if (_.isEmpty(schema.properties)) return;

  return schema;
}

export function parseResponseSchema(
  data:
    | API.PathItemObject['get']
    | API.PathItemObject['put']
    | API.PathItemObject['post']
    | API.PathItemObject['delete']
): Schema | undefined {
  const response = data?.responses?.['200'] as API.ResponseObject;
  const jsonResponse = response.content?.['application/json'];

  if (jsonResponse) {
    // 尝试取 schema
    if (jsonResponse.schema) return jsonResponse.schema as Schema;

    const example =
      jsonResponse.example || (jsonResponse.examples && Object.values(jsonResponse.examples).pop());

    // 尝试转换 example
    if (example) return data2Schema(example);
  }
}

export type IRequestMeta = {
  query?: {
    schema: Schema;
    tsTypeLiteral: string;
  };
  response?: {
    schema: Schema;
    tsTypeLiteral: string;
  };
  pathName: string;
  method: string;
  description?: string;
  tags?: string[];
  apiDoc?:
    | API.PathItemObject['get']
    | API.PathItemObject['put']
    | API.PathItemObject['post']
    | API.PathItemObject['delete'];
};

export interface IServiceMap {
  [serviceName: string]: {
    [requestName: string]: IRequestMeta;
  };
}

export function renderService(apiDoc: API.Document, opt: { operationIdSplitter?: string } = {}) {
  const serviceMap: IServiceMap = {};
  const { operationIdSplitter = '.' } = opt;

  Object.entries(apiDoc.paths).forEach(([pathName, pi]) => {
    if (!pi) return;

    Object.entries({
      GET: pi.get,
      PUT: pi.put,
      POST: pi.post,
      DELETE: pi.delete,
    }).forEach(([method, commonPiData]) => {
      if (!commonPiData) return;

      let [serviceName, requestName] = (
        commonPiData.operationId || `default${operationIdSplitter}${_.camelCase(pathName)}`
      ).split(operationIdSplitter);

      serviceName = _.capitalize(serviceName);

      const requestSchema = parseRequestSchema(commonPiData);
      const respSchema = parseResponseSchema(commonPiData);

      const requestMeta: IRequestMeta = {
        query: requestSchema
          ? { schema: requestSchema, tsTypeLiteral: schema2TsTypeLiteral(requestSchema) }
          : undefined,
        response: respSchema
          ? { schema: respSchema, tsTypeLiteral: schema2TsTypeLiteral(respSchema) }
          : undefined,
        method,
        pathName,
        description: commonPiData.description,
        tags: commonPiData.tags,
        apiDoc: commonPiData,
      };

      //
      _.set(serviceMap, [serviceName, requestName], requestMeta);
    });
  });

  return serviceMap;
}

/** 从数据结构猜测 schema */
export function data2Schema(data: any): Schema {
  if (_.isString(data)) return { type: 'string' };
  if (_.isInteger(data)) return { type: 'integer' };
  if (_.isNumber(data)) return { type: 'number' };

  if (_.isBoolean(data)) return { type: 'boolean' };

  if (_.isArray(data)) {
    return {
      type: 'array',
      items: data2Schema(data[0]),
    };
  }

  if (_.isPlainObject(data)) {
    return {
      type: 'object',
      properties: _.mapValues(data, data2Schema),
      required: Object.keys(data),
    };
  }

  return { type: 'object' };
}
