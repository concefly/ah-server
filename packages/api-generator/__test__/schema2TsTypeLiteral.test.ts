import { schema2TsTypeLiteral } from '../src';

describe('schema2TsTypeLiteral', () => {
  it('oneOf', () => {
    const tsType = schema2TsTypeLiteral({
      oneOf: [
        { type: 'object', properties: { a: { type: 'string' } } },
        { type: 'object', properties: { b: { type: 'string' } } },
      ],
    });

    expect(tsType).toMatchSnapshot();
  });

  it('oneOf deep', () => {
    const tsType = schema2TsTypeLiteral({
      oneOf: [
        { type: 'object', properties: { a: { type: 'string' } } },
        {
          type: 'object',
          properties: {
            b: { type: 'string' },
            c: {
              oneOf: [
                {
                  type: 'object',
                  properties: {
                    cs1: { type: 'string' },
                  },
                  required: ['cs1'],
                },
                {
                  type: 'object',
                  properties: {
                    cs2: { type: 'string' },
                  },
                  required: ['cs2'],
                },
              ],
            },
          },
        },
      ],
    });

    expect(tsType).toMatchSnapshot();
  });

  it('string enum', () => {
    const tsType = schema2TsTypeLiteral({
      type: 'string',
      enum: ['a', 'b', 'c'],
    });

    expect(tsType).toMatchSnapshot();
  });

  it('integer enum', () => {
    const tsType = schema2TsTypeLiteral({
      type: 'integer',
      enum: [1, 2, 3],
    });

    expect(tsType).toMatchSnapshot();
  });
});
