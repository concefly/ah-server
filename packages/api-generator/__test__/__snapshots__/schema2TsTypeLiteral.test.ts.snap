// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema2TsTypeLiteral integer enum 1`] = `"1 | 2 | 3"`;

exports[`schema2TsTypeLiteral oneOf 1`] = `"{ a?: string } | { b?: string }"`;

exports[`schema2TsTypeLiteral oneOf deep 1`] = `"{ a?: string } | { b?: string,c?: { cs1: string } | { cs2: string } }"`;

exports[`schema2TsTypeLiteral string enum 1`] = `"\\"a\\" | \\"b\\" | \\"c\\""`;
